daft/catalog/__init__.py

```diff
@@ -329,10 +329,25 @@ def _from_obj(obj: object) -> Catalog:
     ###
 
     @abstractmethod
-    def create_namespace(self, identifier: Identifier | str): ...
+    def create_namespace(self, identifier: Identifier | str):
+        """Creates a namespace in this catalog."""
+
+    def create_namespace_if_not_exists(self, identifier: Identifier | str):
+        """Creates a namespace in this catalog if it does not already exist."""
+        if not self.has_namespace(identifier):
+            self.create_namespace(identifier)
 
     @abstractmethod
-    def create_table(self, identifier: Identifier | str, source: TableSource) -> Table: ...
+    def create_table(self, identifier: Identifier | str, source: TableSource) -> Table:
+        """Creates a table in this catalog."""
+
+    def create_table_if_not_exists(self, identifier: Identifier | str, source: TableSource) -> Table:
+        """Creates a table in this catalog if it does not already exist."""
+        try:
+            if table := self.get_table(identifier):
+                return table
+        except NotFoundError:
+            return self.create_table(identifier, source)
```

Comment:
should we add this in to the rust session instead of python only? I'd think we would want this functionality in sql too?

Reply:
Agreed long term, but since we don't have SQL create table, we can deliver this now then lower to the rust session later. I believe it would be a lot more work to lower create_table to rust as pre-req to this functionality.